### PR TITLE
fix(cloudprovider/externalgrpc): `GetOptions` returns `MaxNodeProvisionTime` as 0 seconds

### LINKED ISSUE #5935: When using externalgrpc provider, nodes are created and then immediately deleted next loop iteration
**Which component are you using?**:  
Cluster-Autoscaler `externalgrpc` provider

**What version of the component are you using?**: 
`registry.k8s.io/autoscaling/cluster-autoscaler:v1.27.2`

**Component version**: 
helm chart version 9.29.x (edited to allow externalgrpc to consume cloudConfig path)

**What k8s version are you using (`kubectl version`)?**:

<details><summary><code>kubectl version</code> Output</summary><br><pre>
$ kubectl version
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"27", GitVersion:"v1.27.2", GitCommit:"7f6f68fdabc4df88cfea2dcf9a19b2b830f1e647", GitTreeState:"clean", BuildDate:"2023-05-17T14:13:27Z", GoVersion:"go1.20.4", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v5.0.1
Server Version: version.Info{Major:"1", Minor:"26", GitVersion:"v1.26.5+k3s1", GitCommit:"7cefebeaac7dbdd0bfec131ea7a43a45cb125354", GitTreeState:"clean", BuildDate:"2023-05-27T00:05:40Z", GoVersion:"go1.19.9", Compiler:"gc", Platform:"linux/amd64"}

</pre></details>

**What environment is this in?**:
Lab environment

**What did you expect to happen?**:
I'm working on a libvirt module for the externalgrpc provider.  After creating a node via the externalgrpc provisioner, at the next iteration loop the node is then deleted by the autoscaler saying that Scale-Up has timed out.

**What happened instead?**:
I expected the node to be given enough time to provision and register to the kubernetes master (60-90 seconds)

**How to reproduce it (as minimally and precisely as possible)**:
Write a grpc server that creates nodes and then observe the Scale-Up timeout occurring at the next iteration.

**Anything else we need to know?**:
I spent a lot of time in the Kubernetes #sig-autoscaling slack channel discussing this with @vadasambar .  They indicated that this appears to be an issue in the protobuf definition for externalgrpc, as PR #5649 had moved MaxNodeProvisionTime into NodeGroupAutoscalingOptions, but the externalgrpc/protobuf code was not updated to match this change.  They mentioned they will likely be submitting a PR to fix this issue within the next day or so.

### PR DESCRIPTION
**Pull request by vadasambar**
Signed-off-by: vadasambar <surajrbanakar@gmail.com>

#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:
* Adds missing `MaxNodeProvisionTime` in externalgrpc proto
* Sets default `MaxNodeProvisionTime` for nodegroup's `GetOptions` function
* Check #5935 for more info
#### Which issue(s) this PR fixes:

Fixes #5935

#### Does this PR introduce a user-facing change?

```release-note
Fixed: cluster-autoscaler thinks newly scaled up nodegroup using externalgrpc provider has `MaxNodeProvisionTime` set as 0 seconds and expects the new node to be registered in 0-10 seconds instead of the default 15m. Check https://github.com/kubernetes/autoscaler/issues/5935 for more info.  
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: cluster-autoscaler/cloudprovider/externalgrpc/README.md [hunk 1]
```diff
@@ -56,8 +56,8 @@ To regenerate the gRPC code:
 1. install `protoc` and `protoc-gen-go-grpc`:
 
 ```bash
-go install google.golang.org/protobuf/cmd/protoc-gen-go@v1.26
-go install google.golang.org/grpc/cmd/protoc-gen-go-grpc@v1.1
+go install google.golang.org/protobuf/cmd/protoc-gen-go@v1.28
```
Comment (vadasambar) [inline cluster-autoscaler/cloudprovider/externalgrpc/README.md:59]: Updated to latest version at the time of creating this PR. 
image
https://grpc.io/docs/languages/go/quickstart/

File: cluster-autoscaler/cloudprovider/externalgrpc/protos/externalgrpc.pb.go [hunk 2]
```diff
@@ -16,17 +16,17 @@ limitations under the License.
 
 // Code generated by protoc-gen-go. DO NOT EDIT.
 // versions:
-// 	protoc-gen-go v1.26.0
-// 	protoc        v3.19.1
+// 	protoc-gen-go v1.28.1
```
Comment (vadasambar) [inline cluster-autoscaler/cloudprovider/externalgrpc/protos/externalgrpc.pb.go:19]: Everything in this file is auto-generated using 
```
protoc \
  -I ./cluster-autoscaler \
  -I ./cluster-autoscaler/vendor \
  --go_out=. \
  --go-grpc_out=. \
  ./cluster-autoscaler/cloudprovider/externalgrpc/protos/externalgrpc.proto
```
https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler/cloudprovider/externalgrpc#code-generation

File: cluster-autoscaler/cloudprovider/externalgrpc/protos/externalgrpc_grpc.pb.go [hunk 3]
```diff
@@ -15,6 +15,10 @@ limitations under the License.
 */
 
 // Code generated by protoc-gen-go-grpc. DO NOT EDIT.
+// versions:
```
Comment (vadasambar) [inline cluster-autoscaler/cloudprovider/externalgrpc/protos/externalgrpc_grpc.pb.go:18]: Everything in this file is auto-generated using 
```
protoc \
  -I ./cluster-autoscaler \
  -I ./cluster-autoscaler/vendor \
  --go_out=. \
  --go-grpc_out=. \
  ./cluster-autoscaler/cloudprovider/externalgrpc/protos/externalgrpc.proto
```
https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler/cloudprovider/externalgrpc#code-generation

File: cluster-autoscaler/cloudprovider/externalgrpc/protos/externalgrpc.pb.go [hunk 4]
```diff
@@ -786,7 +786,7 @@ type GetAvailableGPUTypesResponse struct {
 	unknownFields protoimpl.UnknownFields
 
 	// GPU types passed in as opaque key-value pairs.
-	GpuTypes map[string]*anypb.Any `protobuf:"bytes,1,rep,name=gpuTypes,proto3" json:"gpuTypes,omitempty" protobuf_key:"bytes,1,opt,name=key,proto3" protobuf_val:"bytes,2,opt,name=value,proto3"`
+	GpuTypes map[string]*any1.Any `protobuf:"bytes,1,rep,name=gpuTypes,proto3" json:"gpuTypes,omitempty" protobuf_key:"bytes,1,opt,name=key,proto3" protobuf_val:"bytes,2,opt,name=value,proto3"`
```
Comment (vadasambar) [inline cluster-autoscaler/cloudprovider/externalgrpc/protos/externalgrpc.pb.go:789]: Note that `any1` is imported from `"github.com/golang/protobuf/ptypes/any"`. `"github.com/golang/protobuf/ptypes/any"` internally uses `anypb`

https://github.com/kubernetes/autoscaler/blob/b569db410fd8bb152c4cc1a940f059c7e2bf0538/cluster-autoscaler/vendor/github.com/golang/protobuf/ptypes/any/any.pb.go#L4-L15

File: cluster-autoscaler/cloudprovider/externalgrpc/externalgrpc_node_group_test.go [hunk 5]
```diff
@@ -248,6 +250,8 @@ func TestCloudProvider_GetOptions(t *testing.T) {
 	assert.Equal(t, 0.7, opts.ScaleDownGpuUtilizationThreshold)
 	assert.Equal(t, time.Minute, opts.ScaleDownUnneededTime)
 	assert.Equal(t, time.Hour, opts.ScaleDownUnreadyTime)
+	assert.Equal(t, time.Hour, opts.ScaleDownUnreadyTime)
```
Comment (dbonfigli) [inline cluster-autoscaler/cloudprovider/externalgrpc/externalgrpc_node_group_test.go:253]: duplicate line

Comment (dbonfigli) [inline cluster-autoscaler/cloudprovider/externalgrpc/externalgrpc_node_group_test.go:253]: duplicated line

**Review by dbonfigli — APPROVED** [review]
Thank you for this.

There is another field that has been recently added to `NodeGroupAutoscalingOptions` , `ZeroOrMaxNodeScaling`, but it is constantly worked on and changed name many times recently, since the default is false and it is not causing issues I would not put it in the protobuf until the feature settles down.

Could you also please update the example at https://github.com/kubernetes/autoscaler/blob/b569db410fd8bb152c4cc1a940f059c7e2bf0538/cluster-autoscaler/cloudprovider/externalgrpc/examples/external-grpc-cloud-provider-service/wrapper/wrapper.go#L349 with the new field?

**Comment by vadasambar** [comment]
@dbonfigli I have addressed the review changes. Thanks for reviewing my PR :pray: 

I guess I need approval from @BigDarkClown to merge the PR.

File: cluster-autoscaler/cloudprovider/externalgrpc/examples/external-grpc-cloud-provider-service/wrapper/wrapper.go [hunk 6]
```diff
@@ -351,6 +351,7 @@ func (w *Wrapper) NodeGroupGetOptions(_ context.Context, req *protos.NodeGroupAu
 		ScaleDownGpuUtilizationThreshold: pbDefaults.GetScaleDownGpuUtilizationThreshold(),
 		ScaleDownUnneededTime:            pbDefaults.GetScaleDownUnneededTime().Duration,
 		ScaleDownUnreadyTime:             pbDefaults.GetScaleDownUnneededTime().Duration,
+		MaxNodeProvisionTime:             pbDefaults.GetMaxNodeProvisionTime().Duration,
```
Comment (dbonfigli) [inline cluster-autoscaler/cloudprovider/externalgrpc/examples/external-grpc-cloud-provider-service/wrapper/wrapper.go:354]: can you please also add the new field in the return value below?

Comment (vadasambar) [reply cluster-autoscaler/cloudprovider/externalgrpc/examples/external-grpc-cloud-provider-service/wrapper/wrapper.go:354]: Added

**Comment by MaciekPytel** [comment]
/approve
I'm going to leave lgtm to @dbonfigli as our general policy is for cloudprovider reviews to be done by their OWNERS and not core CA. The change does look good to me for whatever it's worth.

A separate point - I expect that this will keep happening as more options are moved into GetOptions. And I expect that there will be more fields added there over time - there seem to be a lot of interest in making CA configurable on per-nodegroup level and there isn't really any particular reason for us not to support that. 
With that in mind - is there some long-term solution here? Should we expect anyone adding new fields to GetOptions to extend grpc? If so - can the steps be documented?

Note: I don't think my point above should be blocking for this PR; just a potential follow-up idea.

**Comment by vadasambar** [comment]
@dbonfigli I think this PR needs your `lgtm`.

**Comment by dbonfigli** [comment]
> @dbonfigli I think this PR needs your `lgtm`.

unfortunately i am not a member of the kubernetes org.

**Comment by vadasambar** [comment]
> > @dbonfigli I think this PR needs your `lgtm`.
> 
> unfortunately i am not a member of the kubernetes org.

Thanks for the response. I saw your comment in https://github.com/kubernetes/autoscaler/pull/5998#issuecomment-1702635289

@MaciekPytel we need an LGTM on this from you because @dbonfigli is not a Kubernetes member and cannot add LGTM label to the PR.

**Comment by x13n** [comment]
/lgtm

(based on @dbonfigli's comment above)

**Pull request merged**